storage_controller/src/http.rs

```diff
@@ -785,23 +785,39 @@ async fn handle_tenant_timeline_passthrough(
         error_counter.inc(labels);
     }
 
-    // Transform 404 into 503 if we raced with a migration
-    if resp.status() == reqwest::StatusCode::NOT_FOUND && !consistent {
-        // Rather than retry here, send the client a 503 to prompt a retry: this matches
-        // the pageserver's use of 503, and all clients calling this API should retry on 503.
-        return Err(ApiError::ResourceUnavailable(
-            format!("Pageserver {node} returned 404 due to ongoing migration, retry later").into(),
-        ));
-    }
+    let resp_staus = resp.status();
 
     // We have a reqest::Response, would like a http::Response
-    let mut builder = hyper::Response::builder().status(map_reqwest_hyper_status(resp.status())?);
+    let mut builder = hyper::Response::builder().status(map_reqwest_hyper_status(resp_staus)?);
     for (k, v) in resp.headers() {
         builder = builder.header(k.as_str(), v.as_bytes());
     }
 
+    let resp_bytes = resp
+        .bytes()
+        .await
+        .map_err(|e| ApiError::InternalServerError(e.into()))?;
+
+    // Transform 404 into 503 if we raced with a migration
+    if resp_staus == reqwest::StatusCode::NOT_FOUND && !consistent {
+        let resp_str = std::str::from_utf8(&resp_bytes)
+            .map_err(|e| ApiError::InternalServerError(e.into()))?;
+        // We only handle "tenant not found" errors; other 404s like timeline not found should
+        // be forwarded as-is.
+        if resp_str.contains(&format!("tenant {tenant_or_shard_id}")) {
+            // Rather than retry here, send the client a 503 to prompt a retry: this matches
+            // the pageserver's use of 503, and all clients calling this API should retry on 503.
+            return Err(ApiError::ResourceUnavailable(
+                format!(
+                    "Pageserver {node} returned tenant 404 due to ongoing migration, retry later"
+                )
+                .into(),
+            ));
+        }
+    }
```

Comment:
I wouldn't do this.

Instead, I'd add a `Service::tenant_shard_remote_mutation` version that only operates on shard 0 and use that. Then you should get this behaviour for free.